### PR TITLE
GC/Parallel marking

### PR DESCRIPTION
**Pull request by d-netto**
[As noted below, seems to be causing performance regressions on large number of threads, superseded by https://github.com/JuliaLang/julia/pull/45639.]

This PR extends https://github.com/JuliaLang/julia/pull/41760 by using the deque from https://github.com/JuliaLang/julia/pull/43366 to implement work-stealing in the GC mark loop.

<del> The design is inspired by Horie et al. (https://dl.acm.org/doi/pdf/10.1145/3299706.3210570). 

<del> At a high level, two queues (public/private) are maintained by each thread. The public queue has a fixed size and thieves may steal from it. In case of overflow, elements are pushed into the private queue (which in turn, can be expanded with no need of synchronization, since thieves won't access it).

For the example below (chosen because it spends ~70% of runtime in the mark loop) 

```julia
using BenchmarkTools

gctimes = []

for i in 1:25
    stat = @timed begin
        GC.gc(); GC.gc(); GC.gc()
    end
    push!(gctimes, stat.gctime)
end

@show Threads.nthreads()
@show mean(gctimes)
print("~~~~~~~~")
```

we have

```
for nt in {1..4}; do ../julia/julia -t$nt gc_scrub.jl; done
Threads.nthreads() = 1
mean(gctimes) = 0.35867786796000006
~~~~~~~~
Threads.nthreads() = 2
mean(gctimes) = 0.20352844991999997
~~~~~~~~
Threads.nthreads() = 3
mean(gctimes) = 0.16981833383999997
~~~~~~~~
Threads.nthreads() = 4
mean(gctimes) = 0.14483000495999998
~~~~~~~~
```

The time spent in the mark loop for the example above is artificially large, so these speedups in GC time won't necessarily be achieved in practice.

TODO:

- [x] Fix thread recruitment in `macOS` (current implementation hangs in the `Mach` exception handler).
- [x] Improve heuristics to decide when threads should give up work-stealing or wake up and become thieves (as of now, threads try to steal `jl_n_threads` times and leave the mark loop on failure).
- [ ] Fix GC debugging infrastructure in `src/gc-debug.c`.

### CODE REVIEW

**Comment by chflood** [comment]
I don't have access to the paper unfortunately.  I wish more people used researchgate.net.

My main concern is that work is hidden from the other threads in a private stack and therefore we are missing out on parallelism.  If I pop from my private stack and then generate work, is the new work pushed on the private or public stack?  There is some chance that steals happened and the public stack now has room.

File: src/safepoint.c [hunk 1]
```diff
@@ -156,7 +156,7 @@ void jl_safepoint_end_gc(void)
 
 void jl_safepoint_wait_gc(void)
 {
-    jl_ptls_t ptls = jl_current_task->ptls;    
+    jl_ptls_t ptls = jl_current_task->ptls;
```
Comment (oscardssmith) [inline src/safepoint.c:159]: trailing whitespace.

File: src/julia_threads.h [hunk 2]
```diff
@@ -194,10 +220,11 @@ typedef struct {
     // `1` (atomically). Combining with the sync after marking,
     // this makes sure that a single objects can only appear once in
     // the lists (the mark bit cannot be flipped to `0` without sweeping)
-    void *big_obj[1024];
+    void *big_obj[1024];
```
Comment (vchuravy) [inline src/julia_threads.h:223]:
```suggestion
    void *big_obj[1024];
```

File: src/safepoint.c [hunk 3]
```diff
@@ -155,12 +156,12 @@ void jl_safepoint_end_gc(void)
 
 void jl_safepoint_wait_gc(void)
 {
-    // The thread should have set this is already
-    assert(jl_atomic_load_relaxed(&jl_current_task->ptls->gc_state) != 0);
+    jl_ptls_t ptls = jl_current_task->ptls;
```
Comment (vchuravy) [inline src/safepoint.c:159]:
```suggestion
    jl_ptls_t ptls = jl_current_task->ptls;
```

File: src/signals-mach.c [hunk 4]
```diff
@@ -59,22 +60,7 @@ void jl_mach_gc_end(void)
...
-    }
-    // Otherwise, set the gc state of the thread, suspend and record it
-    int8_t gc_state = ptls2->gc_state;
-    jl_atomic_store_release(&ptls2->gc_state, JL_GC_STATE_WAITING);
-    uintptr_t item = tid | (((uintptr_t)gc_state) << 16);
-    arraylist_push(&suspended_threads, (void*)item);
-    thread_suspend(thread);
-    uv_mutex_unlock(&safepoint_lock);
-    return 1;
+    return jl_atomic_load_relaxed(&jl_gc_running);;
```
Comment (vchuravy) [inline src/signals-mach.c:63]:
```suggestion
    return jl_atomic_load_relaxed(&jl_gc_running);
```

File: src/signals-mach.c [hunk 5]
```diff
@@ -285,7 +284,7 @@ kern_return_t catch_exception_raise(mach_port_t            exception_port,
 #endif
     if (jl_addr_is_safepoint(fault_addr)) {
         if (jl_mach_gc_wait(ptls2, thread, tid))
-            return KERN_SUCCESS;
+            return KERN_FAILURE;
```
Comment (vchuravy) [inline src/signals-mach.c:287]:
```suggestion
            // Fallback to POSIX signals and handle GC recruitment there
            return KERN_FAILURE;
```

**Comment by tveldhui** [comment]
I was digging into long gc times we were seeing (30-40% of duration, gc pauses of 20 seconds, 15 minutes of gc time over a run) and @Sacha0 pointedb out this PR. We will give it a try.

One thing I found might be relevant? Visiting objects in a random memory layout the memory accesses were taking 55ns per object. But if you add `__intrinsic_prefetch(o)` to prefetch the `i+4`th object while marking the i`th` object the time dropped to 16ns per node. I was benchmarking loops like
```
for (int i=0; i < n2; ++i)
    {
       __builtin_prefetch(visit_order[i+4]);
       TreeNode* node = visit_order[i];
       check += node->mark;
       node->mark = 1;
    }
```
not sure if this is something you can make use of. In any case looking forward to trying your parallel marking.

**Comment by d-netto** [comment]
> One thing I found might be relevant? Visiting objects in a random memory layout the memory accesses were taking 55ns per object. But if you add __intrinsic_prefetch(o) to prefetch the i+4th object while marking the ith object the time dropped to 16ns per node.

Yes. Prefetching looks promising for reducing cache misses and it's something that @vchuravy and I are considering as well. IIRC, the OCaml people got some significant performance improvements by prefetching during marking (https://github.com/ocaml/ocaml/pull/10195).

**Comment by vchuravy** [comment]
@nanosoldier `runtests(ALL, vs = ":master")`

**Comment by nanosoldier** [comment]
[Your package evaluation job](https://github.com/JuliaLang/julia/pull/44643#issuecomment-1142115997) has completed - possible new issues were detected. A full report can be found [here](https://s3.amazonaws.com/julialang-reports/nanosoldier/pkgeval/by_hash/7197a7a_vs_f5ebcc9/report.html).

**Review by kpamnany — COMMENTED** [review]
Please either remove the formatting changes or squash your commits such that functional changes are in a separate commit. It will make it easier to review this PR.

**Comment by chriselrod** [comment]
On this PR;
```julia
BenchmarkTools.Trial: 5 samples with 1 evaluation.
 Range (min … max):  45.122 ms … 14.514 s  ┊ GC (min … max): 0.00% … 0.88%
 Time  (median):     97.410 ms             ┊ GC (median):    0.00%
 Time  (mean ± σ):     2.968 s ±  6.455 s  ┊ GC (mean ± σ):  0.86% ± 0.39%

  █
  █▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▆ ▁
  45.1 ms        Histogram: frequency by time         14.5 s <

 Memory estimate: 102.80 MiB, allocs estimate: 576698.
```
Master:
```julia
BenchmarkTools.Trial: 27 samples with 1 evaluation.
 Range (min … max):  104.778 ms … 327.686 ms  ┊ GC (min … max):  0.00% … 50.91%
 Time  (median):     189.883 ms               ┊ GC (median):     0.00%
 Time  (mean ± σ):   187.280 ms ±  52.979 ms  ┊ GC (mean ± σ):  10.68% ± 19.22%

                         █
  ▆▄▁▄▁▁▁▁▄▁▁▁▄▁▁▆▁▁▁▁▄▁▆██▁▄▁▆▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▄▁▁▁▁▁▁▁▁▁▁▄▁▁▁▄ ▁
  105 ms           Histogram: frequency by time          328 ms <

 Memory estimate: 102.80 MiB, allocs estimate: 576672.
```
This is proprietary code, but if needed I could try and make a reproducible open source example.

As is, this PR seems fairly disastrous for performance.

**Comment by chriselrod** [comment]
I now ran with 1000 iterations to get a better idea of the mean; PR:
```julia
julia> extrema(ts), 1e3mean(ts)
((0.037808062, 33.898002256), 466.66967408400006)
```
Master:
julia> extrema(ts), 1e3mean(ts)
((0.03921645, 0.255453254), 59.963547983)
```
7.8x regression on this PR.

**Comment by d-netto** [comment]
@chriselrod, these are some results using the benchmarks and instrumentation from the `GCBenchmarks` repo:

`strings.jl`:

**master:**

1 thread:

```
┌─────────┬────────────┬─────────┬────────────┐
│         │ total time │ gc time │ percent gc │
│         │         ms │      ms │          % │
├─────────┼────────────┼─────────┼────────────┤
│ minimum │      18677 │    4069 │         22 │
│  median │      18863 │    4095 │         22 │
│ maximum │      19041 │    4153 │         22 │
└─────────┴────────────┴─────────┴────────────┘
```

2 threads:

```
┌─────────┬────────────┬─────────┬────────────┐
│         │ total time │ gc time │ percent gc │
│         │         ms │      ms │          % │
├─────────┼────────────┼─────────┼────────────┤
│ minimum │      18724 │    4085 │         21 │
│  median │      19068 │    4130 │         22 │
│ maximum │      19383 │    4199 │         22 │
└─────────┴────────────┴─────────┴────────────┘
```

3 threads:

```
┌─────────┬────────────┬─────────┬────────────┐
│         │ total time │ gc time │ percent gc │
│         │         ms │      ms │          % │
├─────────┼────────────┼─────────┼────────────┤
│ minimum │      18519 │    4058 │         22 │
│  median │      18788 │    4086 │         22 │
│ maximum │      19158 │    4128 │         22 │
└─────────┴────────────┴─────────┴────────────┘
```

**This PR:**

1 thread:

```
┌─────────┬────────────┬─────────┬────────────┐
│         │ total time │ gc time │ percent gc │
│         │         ms │      ms │          % │
├─────────┼────────────┼─────────┼────────────┤
│ minimum │      18833 │    4412 │         23 │
│  median │      18960 │    4449 │         23 │
│ maximum │      19337 │    4555 │         24 │
└─────────┴────────────┴─────────┴────────────┘
```

2 threads:

```
┌─────────┬────────────┬─────────┬────────────┐
│         │ total time │ gc time │ percent gc │
│         │         ms │      ms │          % │
├─────────┼────────────┼─────────┼────────────┤
│ minimum │      18502 │    3809 │         20 │
│  median │      18607 │    3868 │         21 │
│ maximum │      19501 │    3990 │         21 │
└─────────┴────────────┴─────────┴────────────┘
```

3 threads:

```
┌─────────┬────────────┬─────────┬────────────┐
│         │ total time │ gc time │ percent gc │
│         │         ms │      ms │          % │
├─────────┼────────────┼─────────┼────────────┤
│ minimum │      18290 │    3601 │         20 │
│  median │      18431 │    3650 │         20 │
│ maximum │      18718 │    3673 │         20 │
└─────────┴────────────┴─────────┴────────────┘
```

==========

`tree.jl`:

**master:**

1 thread:

```
┌─────────┬────────────┬─────────┬────────────┐
│         │ total time │ gc time │ percent gc │
│         │         ms │      ms │          % │
├─────────┼────────────┼─────────┼────────────┤
│ minimum │       3379 │     313 │          9 │
│  median │       3523 │     323 │          9 │
│ maximum │       3674 │     341 │         10 │
└─────────┴────────────┴─────────┴────────────┘
```

2 threads:

```
┌─────────┬────────────┬─────────┬────────────┐
│         │ total time │ gc time │ percent gc │
│         │         ms │      ms │          % │
├─────────┼────────────┼─────────┼────────────┤
│ minimum │       3433 │     314 │          9 │
│  median │       3549 │     324 │          9 │
│ maximum │       3787 │     347 │         10 │
└─────────┴────────────┴─────────┴────────────┘
```

3 threads:

```
┌─────────┬────────────┬─────────┬────────────┐
│         │ total time │ gc time │ percent gc │
│         │         ms │      ms │          % │
├─────────┼────────────┼─────────┼────────────┤
│ minimum │       3427 │     309 │          9 │
│  median │       3505 │     325 │          9 │
│ maximum │       3635 │     341 │         10 │
└─────────┴────────────┴─────────┴────────────┘
```

**This PR:**

1 thread:

```
┌─────────┬────────────┬─────────┬────────────┐
│         │ total time │ gc time │ percent gc │
│         │         ms │      ms │          % │
├─────────┼────────────┼─────────┼────────────┤
│ minimum │       3447 │     346 │         10 │
│  median │       3509 │     352 │         10 │
│ maximum │       3699 │     367 │         10 │
└─────────┴────────────┴─────────┴────────────┘
```

2 threads:

```
┌─────────┬────────────┬─────────┬────────────┐
│         │ total time │ gc time │ percent gc │
│         │         ms │      ms │          % │
├─────────┼────────────┼─────────┼────────────┤
│ minimum │       3293 │     194 │          6 │
│  median │       3411 │     197 │          6 │
│ maximum │       3557 │     210 │          6 │
└─────────┴────────────┴─────────┴────────────┘
```

3 threads:

```
┌─────────┬────────────┬─────────┬────────────┐
│         │ total time │ gc time │ percent gc │
│         │         ms │      ms │          % │
├─────────┼────────────┼─────────┼────────────┤
│ minimum │       3225 │     142 │          4 │
│  median │       3320 │     148 │          4 │
│ maximum │       3555 │     177 │          5 │
└─────────┴────────────┴─────────┴────────────┘
```

==========

`pollard.jl`:

**master:**

1 thread:

```
┌─────────┬────────────┬─────────┬────────────┐
│         │ total time │ gc time │ percent gc │
│         │         ms │      ms │          % │
├─────────┼────────────┼─────────┼────────────┤
│ minimum │        564 │     143 │         24 │
│  median │        568 │     144 │         25 │
│ maximum │        609 │     152 │         26 │
└─────────┴────────────┴─────────┴────────────┘
```

2 threads:

```
┌─────────┬────────────┬─────────┬────────────┐
│         │ total time │ gc time │ percent gc │
│         │         ms │      ms │          % │
├─────────┼────────────┼─────────┼────────────┤
│ minimum │        566 │     142 │         23 │
│  median │        571 │     144 │         25 │
│ maximum │        621 │     148 │         26 │
└─────────┴────────────┴─────────┴────────────┘
```

3 threads:

```
┌─────────┬────────────┬─────────┬────────────┐
│         │ total time │ gc time │ percent gc │
│         │         ms │      ms │          % │
├─────────┼────────────┼─────────┼────────────┤
│ minimum │        565 │     142 │         23 │
│  median │        570 │     144 │         25 │
│ maximum │        614 │     157 │         27 │
└─────────┴────────────┴─────────┴────────────┘
```

**This PR:**

1 thread:

```
┌─────────┬────────────┬─────────┬────────────┐
│         │ total time │ gc time │ percent gc │
│         │         ms │      ms │          % │
├─────────┼────────────┼─────────┼────────────┤
│ minimum │        574 │     155 │         27 │
│  median │        580 │     157 │         27 │
│ maximum │        590 │     161 │         27 │
└─────────┴────────────┴─────────┴────────────┘
```

2 threads:

```
┌─────────┬────────────┬─────────┬────────────┐
│         │ total time │ gc time │ percent gc │
│         │         ms │      ms │          % │
├─────────┼────────────┼─────────┼────────────┤
│ minimum │        589 │     166 │         27 │
│  median │        594 │     168 │         28 │
│ maximum │        630 │     171 │         29 │
└─────────┴────────────┴─────────┴────────────┘
```

3 threads:

```
┌─────────┬────────────┬─────────┬────────────┐
│         │ total time │ gc time │ percent gc │
│         │         ms │      ms │          % │
├─────────┼────────────┼─────────┼────────────┤
│ minimum │        586 │     162 │         26 │
│  median │        594 │     165 │         28 │
│ maximum │        629 │     168 │         28 │
└─────────┴────────────┴─────────┴────────────┘
```

Indeed, there were performance regressions in some of the benchmarks (e.g. `pollard.jl`), but I didn't observe anything close to that. Could you provide a reproducible open source example? Additionally, any input on the use of `BenchmarkTools` for measuring GC times, @vchuravy?

**Comment by chriselrod** [comment]
Running:
```julia
using DifferentialEquations

function f(du,u,p,t)
  du[1] = p[1] * u[1] - p[2] * u[1]*u[2]
  du[2] = -3 * u[2] + u[1]*u[2]
end

function g(du,u,p,t)
  du[1] = p[3]*u[1]
  du[2] = p[4]*u[2]
end

p = [1.5,1.0,0.1,0.1];
prob = SDEProblem(f,g,[1.0,1.0],(0.0,10.0),p);

function prob_func(prob,i,repeat)
  x = 0.3rand(2)
  remake(prob,p=[p[1:2];x])
end

ensemble_prob = EnsembleProblem(prob,prob_func=prob_func);

include("../../utils.jl")

@gctime solve(ensemble_prob,SRIW1(),trajectories=100_000).u[end].u[end]
```
Most of the time spent is on loading and precompilation. I'll up the trajectories to `1_000_000`.

Running with 36 threads on a 36-thread system yields:
```julia
#  Master:
┌─────────┬────────────┬─────────┬──────────────┬───────────────────┬──────────┬────────────┐
# │         │ total time │ gc time │ max GC pause │ time to safepoint │ max heap │ percent gc │
# │         │         ms │      ms │           ms │                ms │       MB │          % │
# ├─────────┼────────────┼─────────┼──────────────┼───────────────────┼──────────┼────────────┤
# │ minimum │      18032 │    1627 │          817 │                 0 │     2752 │          9 │
# │  median │      18090 │    1699 │          856 │                 0 │     2833 │          9 │
# │ maximum │      19679 │    1894 │          877 │                12 │     2867 │         10 │
# └─────────┴────────────┴─────────┴──────────────┴───────────────────┴──────────┴────────────┘
# ⏎                                                                                                                              
# PR:
# ┌─────────┬────────────┬─────────┬──────────────┬───────────────────┬──────────┬────────────┐
# │         │ total time │ gc time │ max GC pause │ time to safepoint │ max heap │ percent gc │
# │         │         ms │      ms │           ms │                ms │       MB │          % │
# ├─────────┼────────────┼─────────┼──────────────┼───────────────────┼──────────┼────────────┤
# │ minimum │      25287 │    1026 │          739 │               404 │     1210 │          4 │
# │  median │      29303 │    1882 │          906 │               496 │     2950 │          6 │
# │ maximum │      32776 │    2028 │          969 │               690 │     3006 │          7 │
# └─────────┴────────────┴─────────┴──────────────┴───────────────────┴──────────┴────────────┘
```
Note the 50% regression in total time, but claimed gc_time is about the same.

**Comment by chriselrod** [comment]
Reducing the number of threads to 18, matching the number of physical cores, and increasing the number of trajectories to 1_000_000 so that the solve actually spends a decent chunk of the total time:
```julia
# Master:
┌─────────┬────────────┬─────────┬──────────────┬───────────────────┬──────────┬────────────┐
│         │ total time │ gc time │ max GC pause │ time to safepoint │ max heap │ percent gc │
│         │         ms │      ms │           ms │                ms │       MB │          % │
├─────────┼────────────┼─────────┼──────────────┼───────────────────┼──────────┼────────────┤
│ minimum │      44143 │   18857 │         9822 │                 0 │    34017 │         43 │
│  median │      44160 │   19057 │         9919 │                 0 │    34118 │         43 │
│ maximum │      44429 │   19152 │         9948 │                 0 │    34202 │         43 │
└─────────┴────────────┴─────────┴──────────────┴───────────────────┴──────────┴────────────┘

________________________________________________________
Executed in  362.57 secs    fish           external
   usr time   62.67 mins  325.00 micros   62.67 mins
   sys time    3.41 mins   85.00 micros    3.41 mins

# PR:
┌─────────┬────────────┬─────────┬──────────────┬───────────────────┬──────────┬────────────┐
│         │ total time │ gc time │ max GC pause │ time to safepoint │ max heap │ percent gc │
│         │         ms │      ms │           ms │                ms │       MB │          % │
├─────────┼────────────┼─────────┼──────────────┼───────────────────┼──────────┼────────────┤
│ minimum │     171546 │   25212 │        14311 │               324 │    34140 │          4 │
│  median │     200472 │   28749 │        15988 │              4548 │    34569 │         15 │
│ maximum │     567746 │   30408 │        16248 │             31160 │    34794 │         16 │
└─────────┴────────────┴─────────┴──────────────┴───────────────────┴──────────┴────────────┘

________________________________________________________
Executed in   25.31 mins    fish           external
   usr time   68.37 mins  312.00 micros   68.37 mins
   sys time  318.83 mins   84.00 micros  318.83 mins
```
A 3.8 to >12x regression in runtime, but the GC time only increased by about 50%.
Just looking at the GC time vastly understates the degree to which this PR regresses performance.

**Comment by chriselrod** [comment]
Also, on a 64-thread Epyc system, Julia master:
```julia
julia> sim = @time solve(ensemble_prob,SRIW1(),trajectories=10)
 18.253071 seconds (19.22 M allocations: 1.313 GiB, 5.46% gc time, 99.90% compilation time)
EnsembleSolution Solution of length 10 with uType:
RODESolution{Float64, 2, Vector{Vector{Float64}}, Nothing, Nothing, Vector{Float64}, NoiseProcess{Float64, 2, Float64, Vector{Float64}, Vector{Float64}, Vector{Vector{Float64}}, typeof(DiffEqNoiseProcess.INPLACE_WHITE_NOISE_DIST), typeof(DiffEqNoiseProcess.INPLACE_WHITE_NOISE_BRIDGE), true, ResettableStacks.ResettableStack{Tuple{Float64, Vector{Float64}, Vector{Float64}}, true}, ResettableStacks.ResettableStack{Tuple{Float64, Vector{Float64}, Vector{Float64}}, true}, RSWM{Float64}, Nothing, RandomNumbers.Xorshifts.Xoroshiro128Plus}, SDEProblem{Vector{Float64}, Tuple{Float64, Float64}, true, Vector{Float64}, Nothing, SDEFunction{true, typeof(f), typeof(g), UniformScaling{Bool}, Nothing, Nothing, Nothing, Nothing, Nothing, Nothing, Nothing, Nothing, Nothing, Nothing, Nothing, Nothing, typeof(SciMLBase.DEFAULT_OBSERVED), Nothing}, typeof(g), Base.Pairs{Symbol, Union{}, Tuple{}, NamedTuple{(), Tuple{}}}, Nothing}, SRIW1, StochasticDiffEq.LinearInterpolationData{Vector{Vector{Float64}}, Vector{Float64}}, DiffEqBase.DEStats}
```
Versus the PR:
```julia
    PID USER      PR  NI    VIRT    RES    SHR S  %CPU  %MEM     TIME+ COMMAND
1035073 chrisel+  20   0   19.8g 716008 127416 S  4098   0.1   2700:53 julia
```
It's put 2700 CPU minutes in (that's 45 CPU*hours) without finishing yet.

**Comment by chriselrod** [comment]
Also, it is import to test with many threads. That is, with more than four.
With only four threads, the regression is minimal.
Master, running
```julia
julia_master --project=. run_benchmarks.jl -n5 -t4 --bench=binary_tree/tree_immutable.jl
┌─────────┬────────────┬─────────┬──────────────┬───────────────────┬──────────┬────────────┐
│         │ total time │ gc time │ max GC pause │ time to safepoint │ max heap │ percent gc │
│         │         ms │      ms │           ms │                ms │       MB │          % │
├─────────┼────────────┼─────────┼──────────────┼───────────────────┼──────────┼────────────┤
│ minimum │       5977 │    2810 │          127 │                 0 │      459 │         47 │
│  median │       5988 │    2833 │          128 │                 0 │      461 │         47 │
│ maximum │       6122 │    2963 │          130 │                 0 │      462 │         48 │
└─────────┴────────────┴─────────┴──────────────┴───────────────────┴──────────┴────────────┘
```
PR:
```julia
julia_parallel_gc --project=. run_benchmarks.jl -n5 -t4 --bench=binary_tree/tree_immutable.jl
┌─────────┬────────────┬─────────┬──────────────┬───────────────────┬──────────┬────────────┐
│         │ total time │ gc time │ max GC pause │ time to safepoint │ max heap │ percent gc │
│         │         ms │      ms │           ms │                ms │       MB │          % │
├─────────┼────────────┼─────────┼──────────────┼───────────────────┼──────────┼────────────┤
│ minimum │       5828 │    2626 │          103 │                 0 │      462 │         42 │
│  median │       6018 │    2716 │          104 │                 0 │      464 │         46 │
│ maximum │       6247 │    2872 │          106 │                88 │      472 │         46 │
└─────────┴────────────┴─────────┴──────────────┴───────────────────┴──────────┴────────────┘
```
While with 18 threads:
```julia
julia_master --project=. run_benchmarks.jl -n5 -t18 --bench=binary_tree/tree_immutable.jl
┌─────────┬────────────┬─────────┬──────────────┬───────────────────┬──────────┬────────────┐
│         │ total time │ gc time │ max GC pause │ time to safepoint │ max heap │ percent gc │
│         │         ms │      ms │           ms │                ms │       MB │          % │
├─────────┼────────────┼─────────┼──────────────┼───────────────────┼──────────┼────────────┤
│ minimum │       5546 │    3744 │          330 │                 0 │      958 │         67 │
│  median │       5601 │    3793 │          332 │                 0 │      965 │         68 │
│ maximum │       5626 │    3819 │          334 │                 3 │      973 │         68 │
└─────────┴────────────┴─────────┴──────────────┴───────────────────┴──────────┴────────────┘
```
PR:
```julia
julia_parallel_gc --project=. run_benchmarks.jl -n5 -t18 --bench=binary_tree/tree_immutable.jl
┌─────────┬────────────┬─────────┬──────────────┬───────────────────┬──────────┬────────────┐
│         │ total time │ gc time │ max GC pause │ time to safepoint │ max heap │ percent gc │
│         │         ms │      ms │           ms │                ms │       MB │          % │
├─────────┼────────────┼─────────┼──────────────┼───────────────────┼──────────┼────────────┤
│ minimum │       9231 │    5790 │          468 │                 0 │      932 │         48 │
│  median │      10908 │    5809 │          483 │                 4 │      944 │         54 │
│ maximum │      12263 │    6281 │          523 │                71 │      951 │         63 │
└─────────┴────────────┴─────────┴──────────────┴───────────────────┴──────────┴────────────┘
```

**Comment by oscardssmith** [comment]
Chris and I have confirmed (by looking at htop) that he was.

**Comment by chflood** [comment]
I added the start of a deque stats tracker to the stats branch of
https://github.com/chflood/parallelmarking but the numbers aren't adding
up.  I'm new to git so if I did this in the wrong place please let me know.

**Comment by KristofferC** [comment]
> I added the start of a deque stats tracker to the stats branch of https://github.com/chflood/parallelmarking 

As a side note, you would typically not create a completely new repository for this, just add that branch to your already existing fork at https://github.com/chflood/julia.

**Comment by d-netto** [comment]
Superseded by https://github.com/JuliaLang/julia/pull/45639.

**Comment by d-netto** [comment]
Yeah, unfortunately looks like the regressions pointed out above are coming from a live-lock in the thread recruitment scheme.

> I added the start of a deque stats tracker to the stats branch of
https://github.com/chflood/parallelmarking but the numbers aren't adding
up.

It seems that the `num_pops++` in:

```c
STATIC_INLINE void *gc_mark_deque_pop_pc(jl_gc_ws_queue_t *mark_queue)
{
    jl_gc_ws_bottom_t bottom = jl_atomic_load_relaxed(&mark_queue->bottom);
#ifdef ARRAY_AS_PTR
    jl_gc_ws_array_t *array = jl_atomic_load_relaxed(&mark_queue->array);
#else
    jl_gc_ws_array_t *array = &mark_queue->array;
#endif

    int b = bottom.pc_offset - 1;
    jl_gc_ws_bottom_t bottom2 = {b, bottom.data_offset};
    jl_atomic_store_relaxed(&mark_queue->bottom, bottom2);
    jl_fence();
    jl_gc_ws_top_t top = jl_atomic_load_relaxed(&mark_queue->top);
    void *pc;
    if (b >= top.offset) {
        pc = jl_atomic_load_relaxed((_Atomic(void *) *)&array->pc_start[b % array->size]);
        if (__unlikely(b == top.offset)) {
            jl_gc_ws_top_t top2 = {top.offset, top.version + 1};
            if (!jl_atomic_cmpswap(&mark_queue->top, &top, top2)) {
                pc = (void *)_GC_MARK_L_MAX;
                jl_atomic_store_relaxed(&mark_queue->bottom, bottom);
            }
        }
    }
    else {
        pc = (void *)_GC_MARK_L_MAX;
        jl_atomic_store_relaxed(&mark_queue->bottom, bottom);
    }
    deque_stats[jl_threadid()].num_pops++;
    return pc;
}
```

should be move into the `if` statement (otherwise I believe you would be double counting the case where you pop the last element).

I've made a bunch of fixes and superseded this PR with https://github.com/JuliaLang/julia/pull/45639.

**Pull request closed without merging**